lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -4245,6 +4245,151 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
+    if (OpLLVMTy->isVectorTy())
+      IntOpLLVMTy = FixedVectorType::get(IntOpLLVMTy,
+          cast<FixedVectorType>(OpLLVMTy)->getNumElements());
+    SPIRVType *OpSPIRVTy = transType(IntOpLLVMTy);
+    const llvm::fltSemantics &Semantics =
+        OpLLVMTy->getScalarType()->getFltSemantics();
+    const APInt Inf = APFloat::getInf(Semantics).bitcastToAPInt();
+    const APInt AllOneMantissa =
+        APFloat::getLargest(Semantics).bitcastToAPInt() & ~Inf;
+
+    // Integer parameter of the intrinsic is combined from several bit masks
+    // referenced in FPClassTest enum from FloatingPointMode.h in LLVM.
+    // Since a single intrinsic can provide multiple tests - here we might end
+    // up adding several sequences of SPIR-V instructions
+    if (FPClass & fcNan) {
+      auto *TestIsNan = BM->addInstTemplate(OpIsNan, {InputFloat->getId()}, BB,
+                                            ResTy);
+      // Map on OpIsNan if we have both QNan and SNan test bits set
+      if (FPClass & fcSNan && FPClass & fcQNan) {
+        ResultVec.emplace_back(TestIsNan);
+      } else {
+        // isquiet(V) ==> abs(V) >= (unsigned(Inf) | quiet_bit)
+        APInt QNaNBitMask =
+            APInt::getOneBitSet(BitSize, AllOneMantissa.getActiveBits() - 1);
+        APInt InfWithQnanBit = Inf | QNaNBitMask;
+        auto *QNanBitConst = transValue(
+            Constant::getIntegerValue(IntOpLLVMTy, InfWithQnanBit), BB);
+        auto *BitCastToInt = BM->addUnaryInst(OpBitcast, OpSPIRVTy,
+                                              InputFloat, BB);
+        auto *IntAbs =
+            BM->addExtInst(OpSPIRVTy, BM->getExtInstSetId(SPIRVEIS_OpenCL),
+                           OpenCLLIB::SAbs, {BitCastToInt}, BB);
+        auto *TestIsQNan = BM->addCmpInst(OpUGreaterThanEqual, ResTy,
+                                          IntAbs, QNanBitConst, BB);
+        if (FPClass & fcQNan) {
+          ResultVec.emplace_back(TestIsQNan);
+        } else {
+          // issignaling(V) ==> isnan(V) && !isquiet(V)
+          auto *NotQNan =
+              BM->addInstTemplate(OpLogicalNot, {TestIsQNan->getId()}, BB,
+                                  ResTy);
+          auto *TestIsSNan =
+              BM->addInstTemplate(OpLogicalAnd, {TestIsNan->getId(),
+                                  NotQNan->getId()}, BB, ResTy);
+          ResultVec.emplace_back(TestIsSNan);
+        }
+      }
+    }
+    if (FPClass & fcInf) {
+      auto *TestIsInf = BM->addInstTemplate(OpIsInf, {InputFloat->getId()}, BB,
+                                            ResTy);
+      if (FPClass & fcNegInf && FPClass & fcPosInf)
+        // Map on OpIsInf if we have both Inf test bits set
+        ResultVec.emplace_back(TestIsInf);
+      else
+        // Map on OpIsInf with following check for sign bit
+        ResultVec.emplace_back(GetNegPosInstTest(TestIsInf,
+                                                 FPClass & fcNegInf));
```

Comment:
Here and after logical error: check for bit should be done on float, not on IsTest result